### PR TITLE
Implement chain

### PR DESCRIPTION
**Pull request by allouis**
This relates to #1

### CODE REVIEW

**Comment by rjmk** [comment]
Thank you for this (& for your patience)! I am intending to look at it soon.

File: index.js [hunk 1]
```diff
@@ -67,4 +53,3 @@ function of (x) {
     called = true
```
Comment (rjmk) [inline index.js:52]: We need to tweak this function! Your PR threw up the bug. `chain(of, of(1))` loops forever.

Can we have

```js
if (abort) return cb(abort)
var wasCalled = called
called = true
cb(wasCalled, x)
```

Thanks!

**Comment by rjmk** [comment]
It's a bit of a shame that this gives us Applicative + Chain but not Monad. Still, I'm happy with this PR (modulo the comment)

File: index.js (hunk 1 above)
Comment (allouis) [reply index.js:52]: Nice catch!

**Comment by allouis** [comment]
Sorry for late response, only just saw this! Yeah, it's a shame about the lack Monad 😞 I'll keep thinking on it though

**Comment by allouis** [comment]
I've added a fix to `of` and a regression test to go with it ☺️

**Comment by allouis** [comment]
I had a play and double checked the fantasy land spec for Monad, I've got an extra commit showing that this implements the left and right identity for monads. Let me know if i should remove it

File: index.js (hunk 1 above)
Comment (rjmk) [reply index.js:52]: Neat tweak!

File: test.js [hunk 2]
```diff
@@ -12,6 +12,33 @@ assertFull([2,3,4], 'MAP')(source.map(x => x + 1, values([1,2,3])))
...
+assertFull([1,2,3,4], 'CHAIN assoc 2')(source.chain(x => source.chain(source.of, values(x)), values([[1],[2],[3],[4]])))
+
+assertFull([1], 'OF regression')(source.chain(source.of, source.of(1)))
+
+/* Monad */
+
+// left identity
+// M.of(a).chain(f) == f(a)
+
+var f = source.of
```
Comment (rjmk) [inline test.js]: Trying with `f = x => values([x, x, x])` breaks left identity for me

File: test.js [hunk 3]
```diff
@@ -34,4 +61,15 @@ function toArray (cb) {
       read(null, next) //loop again.
     })
   }
-}
\ No newline at end of file
+}
+
+function toArraySync(read){
```
Comment (rjmk) [inline test.js]: I'm not super comfortable with `toArraySync`, though I realise it cleans up the test suite a fair bit

**Comment by rjmk** [comment]
I think we can make it a monad if the chain concats all the results, but then that changes the ap behaviour, and I prefer the current behaviour.

Let me know if I've made a mistake somewhere!

File: test.js (hunk 2 above)
Comment (allouis) [reply test.js]: Ah of course, `chain` only takes first value. grr!

File: test.js (hunk 3 above)
Comment (allouis) [reply test.js]: I've removed this as it was just a helper for left/right identity tests which have been removed

**Comment by allouis** [comment]
@rjmk Yeah I agree, i've removed that last commit. I'm happy with this if you are ☺️

**Comment by rjmk** [comment]
Looks great! Thanks, @allouis :)

**Pull request merged**